### PR TITLE
Reduce `PlaceBuilder` cloning

### PR DESCRIPTION
**Pull request by camsteffen**
Some API tweaks with an eye towards reducing clones.

### CODE REVIEW

File: compiler/rustc_mir_build/src/build/expr/as_place.rs [hunk 1]
```diff
@@ -607,12 +630,12 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
...
-        block = self.bounds_check(block, base_place.clone(), idx, expr_span, source_info);
+        block = self.bounds_check(block, &base_place, idx, expr_span, source_info);
 
         if is_outermost_index {
             self.read_fake_borrows(block, fake_borrow_temps, source_info)
         } else {
-            base_place = base_place.expect_upvars_resolved(self);
+            assert!(base_place.resolve_upvar_mut(self));
             self.add_fake_borrows_of_base(
                 &base_place,
```
Comment (cjgillot) [inline compiler/rustc_mir_build/src/build/expr/as_place.rs:640]: Should this method take a simple `Place`?

File: compiler/rustc_mir_build/src/build/matches/util.rs [hunk 2]
```diff
@@ -97,15 +96,13 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
...
+        mut place: PlaceBuilder<'tcx>,
         pattern: &'pat Pat<'tcx>,
         cx: &Builder<'_, 'tcx>,
     ) -> MatchPair<'pat, 'tcx> {
         // Force the place type to the pattern's type.
         // FIXME(oli-obk): can we use this to simplify slice/array pattern hacks?
-        let mut place = match place.try_upvars_resolved(cx) {
-            Ok(val) | Err(val) => val,
-        };
+        let _is_resolved = place.resolve_upvar_mut(cx);
```
Comment (cjgillot) [inline compiler/rustc_mir_build/src/build/matches/util.rs:105]: Should we call `to_place` directly here?

File: compiler/rustc_mir_build/src/build/expr/as_place.rs (hunk 1 above)
Comment (camsteffen) [reply compiler/rustc_mir_build/src/build/expr/as_place.rs:640]: Yeah that works.

File: compiler/rustc_mir_build/src/build/matches/util.rs (hunk 2 above)
Comment (camsteffen) [reply compiler/rustc_mir_build/src/build/matches/util.rs:105]: No, that panic if the upvar can't be resolved.

**Comment by camsteffen** [comment]
I factored out `resolve_upvar_mut`.

**Comment by cjgillot** [comment]
Great, thanks!
@bors r+

**Pull request merged**